tfsdk/resource_upgrade_state.go

```diff
@@ -0,0 +1,121 @@
+package tfsdk
+
+import (
+	"context"
+
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-go/tfprotov6"
+)
+
+// Optional interface on top of Resource that enables provider control over
+// the UpgradeResourceState RPC. This RPC is automatically called by Terraform
+// when the current Schema type Version field is greater than the stored state.
+// Terraform does not store previous Schema information, so any breaking
+// changes to state data types must be handled by providers.
+//
+// Terraform CLI can execute the UpgradeResourceState RPC even when the prior
+// state version matches the current schema version. The framework will
+// automatically intercept this request and attempt to respond with the
+// existing state. In this situation the framework will not execute any
+// provider defined logic, so declaring it for this version is extraneous.
```

Comment:
Probably not strictly caused by FW, but why "...CLI can execute UpgradeResourceState RPC even when prior state verson matches current schema version"?

And with "extraneous" you mean "pointless" or "superfluous", because the FW will intercept it and so, if a provider hasn't changed schema version, it's dead code?

Reply:
> CLI can execute UpgradeResourceState RPC even when prior state verson matches current schema version

terraform-plugin-sdk used to perform some flatmap -> JSON handling due to changes in Terraform 0.12. Why this was done on the provider side versus the CLI side, I’m not exactly sure. Alas, we must support this odd case until the next major version of the protocol.

> And with "extraneous" you mean "pointless" or "superfluous", because the FW will intercept it and so, if a provider hasn't changed schema version, it's dead code?

Correct, it'll be dead code, similar to declaring state upgraders for versions above the current schema version. I'm just imagining that this might be more of a common gotcha where provider developers might create a state upgrade specifying the current schema version rather than the prior one. Hopefully some unit testing bits can discover that issue and report back to the provider developer about the issue.